text/3396-extern-types-v2.md

```diff
@@ -0,0 +1,249 @@
+}
+```
+The type is `repr(C)` so that it can be used in FFI.
+The `_data` field provides the alignment of the type, and the `_marker` field causes the correct set of automatically implemented traits to be implemented, in this case `!Send`, `!Sync`, `!Unpin`, but `Freeze` (`UnsafeCell` can be used to remove the last one of those).
+This works, however the compiler believes this type is statically sized and aligned which may not be true.
+This means that the type can be easily misused by placing it on the stack, allocating it on the heap, calling `ptr::read`, calling `mem::size_of`, etc.
+This proposal introduces extern types, a way to accurately represent these opaque types in the Rust type system such that the compiler prevents you from misusing it.
+
+The motivation for replacing [RFC 1861] is around computing the offset of fields in aggregate types.
+Extern types do not have an alignment known to the Rust compiler and therefore cannot be included as a field in an aggregate type as it is impossible to correctly calculate their offset, however that RFC did not address this.
+This implies that extern types cannot be included in any (non-`repr(transparent)`) structs and so must be prevented.
+
+Extern types should be able to be used in generic contexts so that they receive blanket trait implementations and can be used inside generic wrapper types.
+Prior to this, all generic types could be included as fields of structs and it was possible to obtain the size and alignment of any type.
+Extern types do not have a size or alignment so this RFC allows users to specify generic bounds sufficiently to statically prevent the above issues.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When writing bindings to foreign code, any types that are opaque to the Rust type system should be declared as follows:
+```rust
+extern {
+    type Foo;
+}
+```
+This is called an extern type and is `!Sized`, `!MetaSized` (explained below), `!Send`, `!Sync`, and is FFI-safe.
+Unlike other dynamically-sized types (DSTs), currently only slices and trait objects, pointers to it (`&Foo`, `&mut Foo`, `*const Foo`, `*mut Foo`, etc) are thin, that is they are 1 `usize` wide not 2.
+
+These types cannot be included in structs as Rust is not able to compute the offset of the field because it does not know the alignment.
+```rust
+struct Header {
+    data: u8,
+    tail: Foo, // Error due to unknown offset
+}
+```
+
+However, these types may be placed in `repr(transparent)` structs:
+```rust
+#[repr(transparent)]
+struct Wrapper<T> {
+    inner: Foo,
+    _marker: PhantomData<T>,
+}
+```
+
+The lack of the `Sized` and `MetaSized` traits on these structs prevent you from calling `ptr::read`, `mem::size_of_val`, etc, which are not meaningful for opaque types.
+
+In the 2021 edition and earlier, these types cannot be used in generic contexts as `T: Sized` and `T: ?Sized` both imply that `T` has a computable size and alignment.
```

Comment:
What if a new-edition trait has an associated type with `?Sized` bound. If now old-edition code is generic over that trait, `Trait::T` would be a not-meta-sized type! IOW, old-edition code might actually have not-meta-sized types in a generic context.

Similarly, what if a new-edition trait has a generic function with `T: ?Sized` bound. If now old-edition code implements this with a `T: ?Sized` bound, this should raise an error since in the old edition,  `T: ?Sized`  actually means  `T: ?Sized+MetaSized`  and this the implementation is less generic than the trait requires.

(Both of these examples came out of boats's recent [exploration](https://without.boats/blog/generic-trait-methods-and-new-auto-traits/) into adding a `Leak` trait.)